### ISSUE TITLE
Guild Bank tabs

### ISSUE THREAD

**Issue by stonedune** [arkayenro/arkinventory#808]
Finally this recent update has forced me to re-configure my bags and try to make them more universal between characters. I was working on my guild bank and can't seem to figure out why if I assign a name/category to bar 5 on tab 1 it also applies it to all there other guild bank tabs.

Is there a setting I'm missing somewhere?

**Comment by joevt** [comment]
All guild bank tabs use the same configuration. That's not new.

What I would like to see is a view that shows all guild tabs at once as a single bag so I don't need to switch between tabs. If you can make mail look like one bag, then doing that for guild bank shouldn't be much more difficult (except mail is take out only). A bar may need an additional option to force items to be stored into a specific set of tabs. That option could be expanded to apply to bags.

**Comment by stonedune** [comment]
I'm ok with the bank tabs being separated, although I do see the allure of that. My screen isn't very big so having 6-7 bars tall (already set 10 wide) starts to cramp my screen and you can't scroll the configure window that I can see.

**Comment by joevt** [comment]
Each ArkInventory window has a Scale setting. I set all mine to 65%. All the windows have a max vertical height setting too (default 2000). If your window is taller than the max height, then ArkInventory adds a vertical scroll bar. Even if the window is too large for your screen, you can drag it by clicking anywhere in the window.

Let's look at the sizes required:

Bags: 136, 10 wide
Bank: 238, 18 wide
Mounts: 308, 20 wide

Mail item: 12*49 = 588, 16 wide. Usually each mail item only has one item instead of 12 (think auction house).

Pets: 975, 35 wide. I rarely use AI to view that. I used 35 to get all pets on the screen at the same time. 35 is only half the width of the screen. I have a rule for each pet type (mechanical, etc). Each pet type has it's own row.

Guild Bank: 686, 16 wide. 784 items if you have 8 tabs.

So guild bank is one of the largest, but not the largest. I think it would be as doable as the Pets and Mail views. I would implement it as a 9th tab probably, or as an entirely new view if there's an inefficiency with having that tab included with the other 8.

**Comment by arkayenro** [comment]
> I was working on my guild bank and can't seem to figure out why if I assign a name/category to bar 5 on tab 1 it also applies it to all there other bar 5 guild bank tabs.

theres only one layout for a window, the guild bank tabs are just bags where you can only see one at a time

> If you can make mail look like one bag, then doing that for guild bank shouldn't be much more difficult

the main issue is that every other location can be scanned as a single list (mail is limited to 50 messages) whenever i want, the vault cannot as its setup completely different to everything else.

the api allows me to set the vault tab, which sends a request for its data, then i have to wait for an event to trigger that tells me the data is ready (it does not state which tab the data is for), i can then scan the data.

vault updates are the same as above, an event to tell you theres been an update and the data is ready for scanning, and the data is always for the current tab, never any of the others (even if another guild member made a change to them)

basically its just too bloody hard to code with the api the way it is

**Comment by joevt** [comment]
When the guild bank is opened, one of the tabs (the currently viewed tab?) has been loaded from the server with QueryGuildBankTab(tab) so that you can use GetGuildBankItemInfo(tab, slot) on those items. What you need to do is query all tabs when the guild bank is opened. I don't think you need to do the query after that (the UI does that for you?) but I'm not sure (I would test with two accounts and see what happens when a item is added or removed to/from the same tab you're viewing and also a different tab). The refresh button could force a rescan if necessary.

The following test code (modified from BigPicture addon) will query all tabs when the bank is opened (ScanGuildBank). The data is scanned (ScanGuildBank2) only after all tabs have been queried (GBFCount > numTabs). The scan will block additional scanning if ScanType = 1 (by setting GBFScanned = 1). Otherwise, scanning will occur after every CHANGE event.

A couple improvements could be made:
1) Skip the query for the default tab when the guild bank is opened (if it is determined that a query actually occurs for the default tab when the guild bank is opened).
2) After all the tabs are scanned the first time, additional scanning should be done on the current tab only (unless it is determined that CHANGE events can occur for tabs that are not the current tab).

This works because the data for previously queried tabs is not removed when you query another tab.

I haven't look at the void storage API so I don't know if something similar could be done for that.

```lua
GBTest = {}
GBTest.items = {}
GBTest.PrevEvent = ""

GBTest.ScanType = 2
-- 1 = scan only at GUILDBANKFRAME_OPENED event
-- 2 = rescan after every GUILDBANKBAGSLOTS_CHANGED event (switching tabs, etc)


function GBTest.ScanGuildBank()
	GBTest.GBFCount = GBTest.GBFCount + 1
	if GBTest.GBFCount <= GetNumGuildBankTabs() then 
		for tab = GBTest.GBFCount, GetNumGuildBankTabs() do
			local _, _, isViewable = GetGuildBankTabInfo(tab)
			GBTest.GBFCount = tab
			if isViewable then
				print("GBTest: Loading Tab "..tab)
				QueryGuildBankTab(tab)
				break
			end
		end
	elseif GBTest.GBFScanned ~= 1 then
		GBTest.ScanGuildBank2()
	end
end


function GBTest.ScanGuildBank2()
	GBTest.items.Guild = {}
	for tab=1, GetNumGuildBankTabs() do
		local scanCount = 0
		local _, _, isViewable = GetGuildBankTabInfo(tab)
		if isViewable then
			GBTest.items.Guild[tab] = {}
			print("GBTest: Scanning Tab "..tab)
			for slot=1,98 do
				local _,itemCount = GetGuildBankItemInfo(tab, slot)
				if itemCount > 0 then
					GBTest.items.Guild[tab][slot] = {}
					local itemLink = GetGuildBankItemLink(tab, slot)
					GBTest.items.Guild[tab][slot].itemLink = itemLink
					GBTest.items.Guild[tab][slot].itemCount = itemCount
				end
			end
		end
	end
	
	if GBTest.ScanType == 1 then
		GBTest.GBFScanned = 1
	else
		GBTest.GBFScanned = nil
	end
end


function GBTest.ScanGuildBank3()
	for tab,t in pairs(GBTest.items.Guild) do
		local scanCount = 0
		for slot,s in pairs(t) do
			scanCount = scanCount + (s.itemCount > 0 and 1 or 0)
		end
		print("GBTest: Tab "..tab.." scanned: "..scanCount)
	end
end


function GBTest.GBHandleEvent(frame, event, ...)
	if event == "GUILDBANKFRAME_OPENED" and event ~= GBTest.PrevEvent then
		GBTest.GBFCount = 0
		GBTest.GBFOpened = 1
		GBTest.items.Guild = {}
		GBTest.ScanGuildBank()
	elseif event == "GUILDBANKBAGSLOTS_CHANGED" then
		if GBTest.GBFOpened == 1 then GBTest.ScanGuildBank() end
	elseif event == "GUILDBANKFRAME_CLOSED" and event ~= GBTest.PrevEvent then
		GBTest.GBFOpened = nil
		GBTest.GBFScanned = nil
		GBTest.ScanGuildBank3()
	end
	GBTest.PrevEvent = event
end


GBTest.GBFrame = CreateFrame("Frame")
GBTest.GBFrame:Show()
GBTest.GBFrame:SetScript("OnEvent", GBTest.GBHandleEvent)
GBTest.GBFrame:RegisterEvent("GUILDBANKFRAME_OPENED")
GBTest.GBFrame:RegisterEvent("GUILDBANKFRAME_CLOSED")
GBTest.GBFrame:RegisterEvent("GUILDBANKBAGSLOTS_CHANGED")
```

**Comment by arkayenro** [comment]
what are you trying to achieve?

previously scanned data is already cached, if you just want one big bag that could probably be done (i'd just stop you from picking up items from 'offline' tabs

all that code above is prone to timing issues, ie was that "tab data ready" event because i asked for it or because the user moved an item, (remember the event doe snot specifiy which tab the data is for so i wont know) it just makes it much more complex for what gain?

**Comment by joevt** [comment]
Right, one big bag is the goal. You explained how the process of scanning the guild bank is difficult. And I replied that there's a reasonably reliable way to do it. It is unlikely that the user could move an item while the guild bank is just opening.

The code is at least guaranteed to do a query of all tabs before scanning the results. There could be a rare occasion where a query is ignored (in case of an extra CHANGE event occurring). It remains to be seen how easy it is to cause such a failure. If it is a problem, then a workaround could be to do an extra query for any tabs where the number of items scanned is zero. It should be a quick query if the number of items was actually zero. If it turns out that items are cached when the guild bank is not opened, then you could skip queries for tabs where the count of items is not zero (meaning they were cached).

Items from offline tabs would be interesting. Can you pick an item from an offline tab? If not, can you switch to the tab and pick it? If not, then can you switch to the offline tab containing the item? That would still make the big view useful, as you could find the item quickly, then click the item to switch to the tab that it's located in.
